### PR TITLE
Reject zero-output linear lowering to TOSA

### PR DESCRIPTION
**Pull request by mukul-sosalekeerthi-ARM**
This rejects matmul-like Torch ops that would produce zero-sized TOSA outputs with a scoped diagnostic, and adds lit/e2e coverage plus TOSA xfail coverage for nn.functional.linear zero-batch input.

### CODE REVIEW

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 1]
```diff
@@ -11406,6 +11436,25 @@ class ConvertTorchToTosa
...
+    bool hasUnsupportedMatmulLikeZeroOutput = false;
+    getOperation()->walk([&](Operation *op) {
+      StringRef opName = op->getName().getStringRef();
+      if (opName != AtenMatmulOp::getOperationName() &&
+          opName != AtenMmOp::getOperationName() &&
+          opName != AtenBmmOp::getOperationName() &&
+          opName != AtenLinearOp::getOperationName() &&
+          opName != "torch.aten.linear")
+        return;
+      if (!hasMatmulLikeZeroOutput(op))
```
Comment (sahas3) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp]: This seems out-of-convention compared to how conversion legality is checked for other ops. Is it not possible to check this condition in the individual rewrite patterns?

**Review by sahas3 — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 1 above)
Comment (mukul-sosalekeerthi-ARM) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp]: Changed this and moved it into the individual conversion patterns.

Comment (sahas3) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp]: Thanks for the change but I had something different in mind. I should have been clearer but I was implying that the check should be inside the _existing_ op rewrite patterns, for example, https://github.com/llvm/torch-mlir/blob/db646b6235da27a65ac58c0dd187b60bbc2f3cf5/lib/Conversion/TorchToTosa/TorchToTosa.cpp#L2346

However, looking at that code I am not sure why the existing check in https://github.com/llvm/torch-mlir/blob/db646b6235da27a65ac58c0dd187b60bbc2f3cf5/lib/Conversion/TorchToTosa/TorchToTosa.cpp#L2386 is not already guarding against zero-dim outputs?

Comment (mukul-sosalekeerthi-ARM) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp]: Thanks, this is now handled in the matmul-family conversion path rather than as a standalone linear special case.

`ConvertAtenLinearOp` inherits from `ConvertAtenMatmulBaseOp`, and `ConvertAtenMatmulBaseOp::checkZeroDimLegality()` now rejects statically zero-sized result tensors before the lowering proceeds. I also added lit coverage for `aten.linear` with a zero-sized output.

**Comment by mukul-sosalekeerthi-ARM** [comment]
@sahas3 Could you please take a look when you get a chance? Thanks.

**Comment by mukul-sosalekeerthi-ARM** [comment]
Hi @sahas3, gentle reminder to please review the changes when you get a chance, Thanks.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp
Comment (sahas3) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:848]: Hi @mukul-sosalekeerthi-ARM, can you please elaborate why is this zero dim check on the output types not already guarding against the scenario where matmul has an output with a zero-dimension?

IIUC, you are only trying to reject matmuls like `[0,4] x [3,4] → [0,3]` which should be caught by this output type check?

For inputs, I understand that `canHandleZeroDimInputs` returns true for matmuls but the output check should still be invalid in this case.

Comment (mukul-sosalekeerthi-ARM) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:845]: Good point. I have removed the matmul specific zero output validation and now rely on the common output zero-dim check.

The only ordering issue was that input zero-dim validation runs before output validation. For cases like [0,4] x [4,3] -> [0,3], the input and output both have a zero dimension, so canHandleZeroDimInputs needs to let the op pass the input precheck and then the existing common output check rejects it. 
The valid zero-contraction case, [5,0] x [0,10] -> [5,10], still works because the output shape has no zero dimension.

**Review by mukul-sosalekeerthi-ARM — COMMENTED** [review]
@sahas3 Removed the expected diagnostic for the deleted matmul-specific check.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 2]
```diff
@@ -851,6 +854,13 @@ class TorchToTosaOpConversionPattern : public OpConversionPattern<AtenOpT> {
   }
 
 protected:
+  virtual LogicalResult
+  checkZeroDimLegality(AtenOpT op, OpAdaptor adaptor,
```
Comment (sahas3) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp]: I think this is not needed anymore? Or is a file missing from this PR that overrides this to `false`?

**Review by sahas3 — APPROVED** [review]
LGTM, assuming the `checkZeroDimLegality` function is not actually needed and will be removed before merging this PR.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 2 above)
Comment (mukul-sosalekeerthi-ARM) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp]: Good catch. I removed the hook and its call, so the base conversion now goes directly through the shared input/output zero-dim checks
